src/main/java/org/opensearch/neuralsearch/processor/normalization/ZScoreNormalizationTechnique.java

```diff
@@ -0,0 +1,234 @@
+    }
+
+    static private float[] findStdPerSubquery(
+        final List<CompoundTopDocs> queryTopDocs,
+        final float[] meanPerSubQuery,
+        final long[] elementsPerSubquery,
+        final int numOfScores
+    ) {
+        final double[] deltaSumPerSubquery = new double[numOfScores];
+        Arrays.fill(deltaSumPerSubquery, 0);
+        for (CompoundTopDocs compoundQueryTopDocs : queryTopDocs) {
+            if (Objects.isNull(compoundQueryTopDocs)) {
+                continue;
+            }
+            List<TopDocs> topDocsPerSubQuery = compoundQueryTopDocs.getTopDocs();
+            int subQueryIndex = 0;
+            for (TopDocs topDocs : topDocsPerSubQuery) {
+                for (ScoreDoc scoreDoc : topDocs.scoreDocs) {
+                    deltaSumPerSubquery[subQueryIndex] += Math.pow(scoreDoc.score - meanPerSubQuery[subQueryIndex], 2);
+                }
+                subQueryIndex++;
+            }
+        }
+
+        final float[] stdPerSubQuery = new float[numOfScores];
+        for (int i = 0; i < deltaSumPerSubquery.length; i++) {
+            if (elementsPerSubquery[i] == 0) {
+                stdPerSubQuery[i] = 0;
+            } else {
+                stdPerSubQuery[i] = (float) Math.sqrt(deltaSumPerSubquery[i] / elementsPerSubquery[i]);
+            }
+        }
+
+        return stdPerSubQuery;
+    }
+
+    static private float sumScoreDocsArray(final ScoreDoc[] scoreDocs) {
+        float sum = 0;
+        for (ScoreDoc scoreDoc : scoreDocs) {
+            sum += scoreDoc.score;
+        }
+
+        return sum;
+    }
+
+    private ZScores getZScoreResults(final List<CompoundTopDocs> queryTopDocs) {
+        int numOfSubqueries = getNumOfSubqueries(queryTopDocs);
+
+        // to be done for each subquery
+        float[] sumPerSubquery = findScoreSumPerSubQuery(queryTopDocs, numOfSubqueries);
+        long[] elementsPerSubquery = findNumberOfElementsPerSubQuery(queryTopDocs, numOfSubqueries);
+        float[] meanPerSubQuery = findMeanPerSubquery(sumPerSubquery, elementsPerSubquery);
+        float[] stdPerSubquery = findStdPerSubquery(queryTopDocs, meanPerSubQuery, elementsPerSubquery, numOfSubqueries);
+        return new ZScores(meanPerSubQuery, stdPerSubquery);
+    }
+
+    private static float normalizeSingleScore(final float score, final float standardDeviation, final float mean) {
+        // edge case when there is only one score and z scores are same
+        if (Floats.compare(mean, score) == 0) {
+            return SINGLE_RESULT_SCORE;
+        }
+        // Case when sd is 0
+        if (Floats.compare(standardDeviation, 0.0f) == 0) {
+            return MIN_SCORE;
+        }
+        float normalizedScore = (score - mean) / standardDeviation;
+
+        return normalizedScore <= 0.0f ? MIN_SCORE : normalizedScore;
+    }
+
+    /**
+     * Result class to hold mean and std dev scores for each sub query
+     */
+    @AllArgsConstructor
+    @Getter
+    private class ZScores {
```

Comment:
with JDK 21 you can now use record key word with classes like this. Try it out. Ref: https://www.baeldung.com/java-record-keyword